src/freenet/clients/fcp/FCPPluginClient.java

```diff
@@ -0,0 +1,1190 @@
+        // We do this through the Map FCPPluginClient.synchronousSends, which is guarded by.
+        // FCPPluginClient.synchronousSendsLock. Also see the JavaDoc of the Map for an overview of
+        // this mechanism.
+        
+        if(!message.isReplyMessage()) {
+            return false;
+        }
+
+        // Since the JavaDoc of sendSynchronous() tells people to use it not very often due to
+        // the impact upon thread count, we assume that the percentage of messages which pass
+        // through here for which there is an actual sendSynchronous() thread waiting is small.
+        // Thus, a ReadWriteLock is used, and we here only take the ReadLock, which can be taken
+        // by *multiple* threads at once. We then read the map to check whether there is a
+        //  waiter, and if there is, take the write lock to hand the message to it.
+        // (The implementation of ReentrantReadWritelock does not allow upgrading a readLock()
+        // to a writeLock(), so we must release it in between and re-check afterwards.)
+
+        synchronousSendsLock.readLock().lock();
+        try {
+            if(!synchronousSends.containsKey(message.identifier)) {
+                return false;
+            }
+        } finally {
+            synchronousSendsLock.readLock().unlock();
+        }
+        
+        synchronousSendsLock.writeLock().lock();
+        try {
+            SynchronousSend synchronousSend = synchronousSends.get(message.identifier);
+            if(synchronousSend == null) {
+                // The waiting sendSynchronous() has probably returned already because its
+                // timeout expired.
+                // So by returning false, we ask the caller to deliver the message to the
+                // regular message handling interface to make sure that it is not lost.
+                return false;
+            }
+
+            assert(synchronousSend.reply == null)
+                : "One identifier should not be used for multiple messages or replies";
+
+            synchronousSend.reply = message;
+            // Wake up the waiting synchronousSend() thread
+            synchronousSend.completionSignal.signal();
+
+            return true;
+        } finally {
+            synchronousSendsLock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * Backend for {@link #send(SendDirection, FCPPluginMessage)} to dispatch messages to a
+     * {@link FredPluginFCPMessageHandler}.<br><br>
+     * 
+     * This shall only be called for messages for which it was determined that the message handler
+     * is a plugin running in the local VM.<br><br>
+     * 
+     * The message will be dispatched in a separate thread so this function can return quickly.
+     */
+    private void dispatchMessageLocallyToMessageHandler(
+            final FredPluginFCPMessageHandler messageHandler, final SendDirection direction,
+            final FCPPluginMessage message) {
```

Comment:
>Most of this method is message handling logic. Why is this implemented here instead of making the FredPluginFCPMessageHandler responsible for this?

FredPluginFCPMessageHandler is an interface, not an abstract class, so it cannot contain any code. And I would like to keep it that way, i.e. not make it abstract and add code, to ensure that client/server authors have a simple, short thing which they can read to understand message handling. It shall only contain stuff interesting to the server/client, not internal implementation details. Arguable its member class FCPPluginMessage violates that, but it shall be moved outside before the pull request is merged I suppose. Dunno whether someone complained about that already, but I guess I shall do it nevertheless.
(Another reason for it not being an abstract class but an interface is that it is necessary due to the way plugins are loaded: Their main class shall implement all relevant FredPlugin... interfaces. Having multiple would not be possible if they were abstract, as Java does not allow multiple inheritance, only multiple interface implementations. It would also be a severe restriction upon plugin authors to force them to extend this specific class.)
ACK?

> The getPriority() implementation in the Runnable ìn this method could then also be far prettier (i.e. without instanceof magic).

Hmm. What do you suggest instead?
Also, why are you against instanceof? IMHO optionally implementing an interface is a beautiful way, if not the de-facto way, of allowing code to add  optional functions. getPriority() being optional is a good example. If you're against it for performance reasons: Google "instanceof performance" yields several stackexchange.com threads which you might check for proof whether it is actually slow enough to justify more ugly code. The class hierarchy might be cached by the JVM very nicely... Have not checked myself though.